File: drimtajm/day1/day1.rb
require 'benchmark'
modules = File.read("input1.txt").split.map(&:to_i)

def calculate_fuel(mass)
  fuel = mass.div(3) - 2
  (fuel > 0 ? fuel : 0)
end

def calculate_total_fuel(modules)
  total_fuel = 0
  for f in modules
    while f > 0
      f = calculate_fuel(f)
      total_fuel += f
    end
  end
  total_fuel
end

time = Benchmark.measure {
  puts calculate_total_fuel modules
}
puts time.real

File: drimtajm/day1/day1.erl
-module('day1').
-compile('export_all').

read_numbers(FileName) ->
    {ok, Device} = file:open(FileName, [read]),
    try get_all_numbers(Device, [])
      after file:close(Device)
    end.

get_all_numbers(Device, Numbers) ->
    case file:read_line(Device) of
        eof        ->
	    lists:reverse(Numbers);
        {ok, Data} ->
	    Number = list_to_integer(string:strip(Data, right, $\n)),
	    get_all_numbers(Device, [Number | Numbers])
    end.

calculate_fuel(Modules) ->
    calculate_fuel(Modules, []).
calculate_fuel([], FuelValues) ->
    lists:reverse(FuelValues);
calculate_fuel([Mass | Rest], FuelValues) ->
    Fuel = (Mass div 3) - 2,
    if (Fuel > 0) -> calculate_fuel(Rest, [Fuel | FuelValues]);
       true       -> calculate_fuel(Rest, FuelValues)
    end.

calculate_total_fuel(Modules) ->
    calculate_total_fuel(Modules, 0).
calculate_total_fuel([], Fuel) ->
    Fuel;
calculate_total_fuel(Modules, Fuel) ->
    FuelValues = calculate_fuel(Modules),
    calculate_total_fuel(FuelValues, Fuel+lists:sum(FuelValues)).

go() ->
    Numbers = read_numbers("input1.txt"),
    calculate_total_fuel(Numbers).
